ergotree-ir/src/sigma_protocol/sigma_boolean.rs

```diff
@@ -196,6 +211,22 @@ mod arbitrary {
         }
     }
 
+    impl Arbitrary for ProveDhTuple {
+        type Parameters = ();
+        type Strategy = BoxedStrategy<Self>;
+
+        fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+            (
+                any::<EcPoint>(),
+                any::<EcPoint>(),
+                any::<EcPoint>(),
+                any::<EcPoint>(),
+            )
+                .prop_map(|(gv, hv, uv, vv)| ProveDhTuple::new(gv, hv, uv, vv))
+                .boxed()
+        }
+    }
+
     impl Arbitrary for SigmaBoolean {
```

Comment:
There was an existing SigmaBoolean Arbitrary impl here that uses the ProveDlog enum variant.  
Do we implementations for the other variants (ProveDHTuple, Cand, Cor, Cthreshold)

Reply:
Good question! Please, extend `SigmaBoolean` impl with DH tuple `Cand` and `Cor` conjectures and adding `Arbitrary` impl for them along the way. 
Be aware that `SigmaBoolean` is extensively used in PBT tests (through `SigmaProp`) for boxes, transaction building and signing, etc. Watch for the impact on overall test suite run time. We need to keep it reasonable. :)

Reply:
Would something like this work?   
`    
        fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
            prop_oneof![
                (any::<ProveDlog>()),
                (any::<ProveDhTuple>()),
                (any::<Cand>()),
                (any::<Cor>()),
                (any::<Cthreshold>()),
            ]
            .prop_map(|p| match p {
                SigmaProofOfKnowledgeTree::ProveDlog(provedlog) => {
                    SigmaBoolean::ProofOfKnowledge(SigmaProofOfKnowledgeTree::ProveDlog(provedlog))
                }
                SigmaProofOfKnowledgeTree::ProveDhTuple(provedhtuple) => {
                    SigmaBoolean::ProofOfKnowledge(SigmaProofOfKnowledgeTree::ProveDhTuple(
                        provedhtuple,
                    ))
                }
                SigmaConjecture::Cand(cand) => {
                    SigmaBoolean::SigmaConjecture(SigmaConjecture::Cand(cand))
                }
                SigmaConjecture::Cor(cor) => {
                    SigmaBoolean::SigmaConjecture(SigmaConjecture::Cor(cor))
                }
                SigmaConjecture::Cthreshold(cthreshold) => {
                    SigmaBoolean::SigmaConjecture(SigmaConjecture::Cthreshold(cthreshold))
                }
            })
            .boxed()
        }
    }
`

Also, I am still working on the `Arbitrary` impl for `Cand`/`Cor`...it seems I need to bound the `SigmaConjectureItems` vectors between 2 and 1000?  Maybe its better to keep the size small to minimize test time as you mentioned?

Reply:
1. There is a `prop_recursive` for the generation of recursive structures. See how it's used in constant generator - https://github.com/ergoplatform/sigma-rust/blob/e4965fcd758c0d154bb81ec424d4c70a189a8514/ergotree-ir/src/mir/constant.rs#L313-L359 . 
2. There should be `From` implemented for `SigmaBoolean` so that would allow simple `.into()` or `.prop_map_into()`.
3. Yes, something like `2..5` should suffice for conjecture items.